onnx/defs/shape_inference.h

```diff
@@ -44,8 +44,22 @@ inline bool hasExactlyNInputTypes(InferenceContext& ctx, int n, const std::strin
   return true;
 }
 
+inline bool hasNInputShapes(InferenceContext& ctx, int n, const std::string& opname) {
+  if (static_cast<int>(ctx.getNumInputs()) < n) {
+    throw std::runtime_error(opname + " has too few inputs");
+  }
+  for (int i = 0; i < n; i++) {
+    if (!ctx.getInputType(i) || !ctx.getInputType(i)->has_shape()) {
+      return false;
+    }
+  }
+  return true;
+}
+
 inline void propagateElemTypeFromInputToOutput(InferenceContext& ctx, size_t inputIndex, size_t outputIndex) {
-  ctx.getOutputType(outputIndex)->set_elem_type(ctx.getInputType(inputIndex)->elem_type());
+  if (ctx.getInputType(inputIndex)) {
```

Comment:
maybe assert inputIndex and outputIndex are not out of bound?

Reply:
actually it might be better to assert inside getInputType and getOutputType

Reply:
done, see implementation.h